Fixing bug where fields could be prefixed with table name.

nsheadlight opened the pull request:
If a join is present with that table name look to find table name qualifiers on the field.


This is a fix for the following example query where the field name is qualified by the table name:

`/Documents/FilteredTo/FBJL~Project.CustomProperties.parish~INN~01/0/10?Joins=Project,IDProject,IDProject`

Normally this qualification isnt an issue but because JSON fields extract the sub-path it incorrectly extracts it.

nsheadlight committed:
Fixing bug where fields could be prefixed with table name. If a join is present with that table name look to find table name qualifiers on the field

bigal-hl commented on source/Meadow-Filter-Parse.js:
```diff
@@ -283,37 +296,37 @@ const addFilterStanzaToQuery = (pFilterStanza, pQuery) =>
 		case 'FBJV':   // Filter by JSON Value (left-side AND)
 		{
-			const { fieldColumn, jsonPath } = parseJSONFieldAndPath(pFilterStanza);
+			const { fieldColumn, jsonPath } = parseJSONFieldAndPath(pFilterStanza, pQuery.parameters?.join);
```
This `?` syntax puts a minimum node 14 requirement on the user. Not sure if that is safe for this package which is used by anything using 2.x meadow endpoints. Guessing this will break some older consumers, so lets just make this use `pQuery.parameters && pQuery.parameters.join` for compatibility.

This also adds two dependencies I don't love:

* It makes this library care about more query state than it previously did. Before, all it ever did was add a where clause. I didn't inspect the state of the query at all.
* It cares about the sequencing that the query is mutated. If the joins are added later on, this won't do what we expect.

We can maybe live with #1, but for #2, does this change actually do what you want? I am not sure I know the exact order that things are processed.

Overall, I'd be much happier if we can make the two syntaxes unambiguous and make this parser not care about the internal state of the query object.

nsheadlight replied:
I agree with both points, it seems like the root of the issue is like you said the ambiguity between the sql and jql part. Maybe something like

```
Field: FormDataJSON$.FormData.Header.Category
Result: JSON_EXTRACT(FormDataJSON, '$.FormData.Header.Category')
```

Which would allow for a clear separation between sql field and json field. The main thing I dont like about that is the syntax but it does mean you get to define what the json path specifically should be

nsheadlight replied:
@bigal-hl updated the PR to use that strategy. LMK what you think

nsheadlight committed:
Updating to address PR review.

Moved to a different syntax of $ in the field name which allows for specifying full jq path and clearly separates field and path

bigal-hl commented on test/test.js:
```diff
@@ -1,12 +1,13 @@
 suite('Filter Stanza Parse', () =>
 {
     const parse = require('../source/Meadow-Filter').parse;
-    let queryStub = { addFilter: () => { }, setDistinct: () => { }, addSort: () => { } };
+    let queryStub = { addFilter: () => { }, setDistinct: () => { }, addSort: () => { }, parameters: { join: [] } };
```
(nit) Can revert.

nsheadlight committed:
Removing old test code not needed from first pass
